File: app/services/scrapbook.ts
import { API_BASE_URL } from '@/constants/config';
import {
  CollectibleCatalogEntry,
  DiscoveryDetails,
  DiscoveryResponse,
  CollectionStats,
  QRCodeInfo,
} from '@/types/scrapbook';

export const scrapbookService = {
  async getCatalog(token: string): Promise<CollectibleCatalogEntry[]> {
    const url = `${API_BASE_URL}/scrapbook/catalog`;
    
    const response = await fetch(url, {
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });
    
    if (!response.ok) {
      throw new Error(`Failed to fetch catalog: ${response.statusText}`);
    }
    
    return response.json();
  },

  async scanQRCode(token: string, qrToken: string): Promise<DiscoveryResponse> {
    const url = `${API_BASE_URL}/scrapbook/scan`;
    
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify({ qr_token: qrToken }),
    });
    
    if (!response.ok) {
      throw new Error(`Failed to scan QR code: ${response.statusText}`);
    }
    
    return response.json();
  },

  async getDiscoveryDetails(token: string, catalogEntryId: number): Promise<DiscoveryDetails> {
    const url = `${API_BASE_URL}/scrapbook/discovery/${catalogEntryId}`;
    
    const response = await fetch(url, {
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });
    
    if (!response.ok) {
      throw new Error(`Failed to fetch discovery: ${response.statusText}`);
    }
    
    return response.json();
  },

  async updateNotes(token: string, discoveryId: number, notes: string): Promise<void> {
    const url = `${API_BASE_URL}/scrapbook/discovery/${discoveryId}/notes`;
    
    const response = await fetch(url, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify({ notes }),
    });
    
    if (!response.ok) {
      throw new Error(`Failed to update notes: ${response.statusText}`);
    }
  },

  async toggleFavorite(token: string, discoveryId: number): Promise<boolean> {
    const url = `${API_BASE_URL}/scrapbook/discovery/${discoveryId}/favorite`;
    
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });
    
    if (!response.ok) {
      throw new Error(`Failed to toggle favorite: ${response.statusText}`);
    }
    
    const data = await response.json();
    return data.is_favorite;
  },

  async getCollectionStats(token: string): Promise<CollectionStats> {
    const url = `${API_BASE_URL}/scrapbook/stats`;
    
    const response = await fetch(url, {
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });
    
    if (!response.ok) {
      throw new Error(`Failed to fetch stats: ${response.statusText}`);
    }
    
    return response.json();
  },

  async getQRInfo(qrToken: string): Promise<QRCodeInfo> {
    const url = `${API_BASE_URL}/scrapbook/qr/${qrToken}`;
    
    const response = await fetch(url);
    
    if (!response.ok) {
      throw new Error(`Failed to fetch QR info: ${response.statusText}`);
    }
    
    return response.json();
  },
};

File: api/routers/scrapbook.py
from fastapi import APIRouter, Depends, HTTPException, status
from typing import List, Optional
from uuid import UUID

from models.scrapbook import (
    CollectibleCatalogEntry,
    DiscoveryDetails,
    DiscoveryResponse,
    CollectionStats,
    UpdateNotesRequest,
    ScanQRRequest,
    QRCodeInfo,
)
from services.scrapbook import ScrapbookService
from dependencies import get_scrapbook_service, get_current_user

router = APIRouter(prefix="/scrapbook", tags=["scrapbook"])


@router.get("/catalog", response_model=List[CollectibleCatalogEntry])
async def get_collectible_catalog(
    user=Depends(get_current_user),
    service: ScrapbookService = Depends(get_scrapbook_service),
):
    """
    Get the full collectible catalog with user's discovery status.
    Shows all plants available to collect and which ones the user has discovered.
    """
    return await service.get_user_catalog(user["id"])


@router.post("/scan", response_model=DiscoveryResponse)
async def scan_qr_code(
    request: ScanQRRequest,
    user=Depends(get_current_user),
    service: ScrapbookService = Depends(get_scrapbook_service),
):
    """
    Scan a QR code to discover a plant.
    Records the discovery in the user's scrapbook if not already discovered.
    """
    return await service.scan_qr_code(user["id"], request.qr_token)


@router.get("/discovery/{catalog_entry_id}", response_model=Optional[DiscoveryDetails])
async def get_discovery_details(
    catalog_entry_id: int,
    user=Depends(get_current_user),
    service: ScrapbookService = Depends(get_scrapbook_service),
):
    """
    Get detailed information about a discovered plant.
    Returns None if the plant hasn't been discovered by the user yet.
    """
    details = await service.get_discovery_details(user["id"], catalog_entry_id)
    if not details:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Discovery not found or plant not yet discovered",
        )
    return details


@router.put("/discovery/{discovery_id}/notes", response_model=dict)
async def update_discovery_notes(
    discovery_id: int,
    request: UpdateNotesRequest,
    user=Depends(get_current_user),
    service: ScrapbookService = Depends(get_scrapbook_service),
):
    """
    Update personal notes for a discovered plant.
    """
    success = await service.update_discovery_notes(
        user["id"], discovery_id, request.notes
    )
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Discovery not found or unauthorized",
        )
    return {"success": True, "message": "Notes updated successfully"}


@router.post("/discovery/{discovery_id}/favorite", response_model=dict)
async def toggle_discovery_favorite(
    discovery_id: int,
    user=Depends(get_current_user),
    service: ScrapbookService = Depends(get_scrapbook_service),
):
    """
    Toggle the favorite status of a discovered plant.
    Returns the new favorite status.
    """
    new_status = await service.toggle_favorite(user["id"], discovery_id)
    return {"success": True, "is_favorite": new_status}


@router.get("/stats", response_model=CollectionStats)
async def get_collection_stats(
    user=Depends(get_current_user),
    service: ScrapbookService = Depends(get_scrapbook_service),
):
    """
    Get statistics about the user's plant collection.
    Includes total counts, percentages, and rarity breakdowns.
    """
    return await service.get_collection_stats(user["id"])


@router.get("/qr/{qr_token}", response_model=QRCodeInfo)
async def get_qr_info(
    qr_token: UUID,
    service: ScrapbookService = Depends(get_scrapbook_service),
):
    """
    Get information about a QR code (public endpoint for previewing before scan).
    """
    qr_info = await service.get_qr_info(qr_token)
    if not qr_info:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="QR code not found",
        )
    return qr_info


@router.post("/qr/create", response_model=QRCodeInfo)
async def create_qr_code(
    species_id: int,
    user=Depends(get_current_user),
    service: ScrapbookService = Depends(get_scrapbook_service),
):
    """
    Create a new QR code for a plant species.
    Admin/staff endpoint for generating QR codes.
    """
    return await service.create_qr_code(species_id)


@router.post("/qr/{qr_code_id}/deactivate", response_model=dict)
async def deactivate_qr_code(
    qr_code_id: int,
    user=Depends(get_current_user),
    service: ScrapbookService = Depends(get_scrapbook_service),
):
    """
    Deactivate a QR code.
    Admin/staff endpoint for managing QR codes.
    """
    success = await service.deactivate_qr_code(qr_code_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="QR code not found",
        )
    return {"success": True, "message": "QR code deactivated"}

